### ISSUE TITLE
Masked decoder use leads to excessive memory usage (mask=mask)

### ISSUE THREAD

**Issue by andyschuster** [lucidrains/x-transformers#325]
The decoder module creates excessive (~4x) memory demands when using masked attention. That increase seems way higher than expected from just adding a mask and might point to a memory leak somewhere. Anyone else experiencing this?

```
import torch
from x_transformers import TransformerWrapper, Decoder

transformer = TransformerWrapper(
    num_tokens=64,
    max_seq_len=20000,
    attn_layers=Decoder(
        dim=64,
        depth=8,
        heads=8,
        attn_flash=True,
    ),
).cuda()

mask = torch.ones(8, 2000).bool().cuda()
tokens = torch.randint(low=0, high=63, size=(8, 2000)).cuda()

# allocates ~2.5 GB GPU memory
transformer(tokens)

# allocates ~10.4 GB GPU memory
transformer(tokens, mask=mask)
```

Versions:
> python 3.12.10 
> torch v2.7.0+cu128
> x-transformers v2.3.5
> 
> Ubuntu 24.04.2 LTS in WSL2

**Comment by lucidrains** [comment]
@andyschuster hey, so i think the right way is to have a (triton) kernel that can accept key padding mask in addition to being causal. surprisingly i don't think it exists yet

but another way would be to constitute the full attention mask before passing through the attention layers (when it detects causal and key padding mask used in conjunction)
